model-optimizer/extensions/back/SpecialNodesFinalization.py

```diff
@@ -126,26 +126,24 @@ class RemoveConstToResult(BackReplacementPattern):
     def pattern():
         return dict(
             nodes=[
-                ('const_node', {'type': 'Const', 'kind': 'op'}),
-                ('const_data', {'kind': 'data'}),
+                ('const_data', {'kind': 'data', 'value': lambda value: value is not None}),
                 ('result_node', {'type': 'Result', 'kind': 'op'}),
             ],
             edges=[
-                ('const_node', 'const_data'),
                 ('const_data', 'result_node')
             ]
         )
 
     @staticmethod
     def replace_pattern(graph: Graph, match: dict):
-        const_node = match['const_node']
         const_data_node = match['const_data']
         result_node = match['result_node']
         nodes_to_remove = [result_node.id]
 
         # in case only const data consumer that is the result node, remove the whole sub-graph
-        if len(const_node.out_port(0).get_destinations()) == 1:
-            nodes_to_remove.append(const_node.id)
+        parent_node = result_node.in_port(0).get_source().node
+        if parent_node.soft_get('type') == 'Const' and len(parent_node.out_port(0).get_destinations()) == 1:
+            nodes_to_remove.append(parent_node.id)
```

Comment:
After this change we may have a situation when the graph is split into 2 unconnected components. Probably we need to make this transformation run graph clean-up.

Reply:
There won't be unconnected components. 

Algorithm:
1. Search for Result with incoming valued data
2. Check if this valued data was caused by Const operation. If so -- Delete Const, data and Result  -- all three of them
3. If valued data is caused by constant sub-graph -- delete Result operation from the graph. The rest of the nodes will be deleted by graph_clean_up -- but resulting graph itself is in a pretty consistent state.

Should we call graph.clean_up anyway?

Reply:
I agree that the graph is in the consistent state. But as you mentioned in the item 3 the part of the constant sub-graph will be disconnected from the main graph after removal of the `parent_node`. I vote for adding `force_cleanup` attribute to the transformation to avoid possible chance that no other transformation call it later.